src/target/linux.rs

```diff
@@ -221,6 +306,192 @@ impl LinuxTarget {
+                ptrace::ptrace(
+                    ptrace::Request::PTRACE_POKEUSER,
+                    self.pid,
+                    (*DEBUG_REG_OFFSET + index * 8) as *mut libc::c_void,
+                    breakpoint.addr as *mut libc::c_void,
+                )?;
+                ptrace::ptrace(
+                    ptrace::Request::PTRACE_POKEUSER,
+                    self.pid,
+                    (*DEBUG_REG_OFFSET + 7 * 8) as *mut libc::c_void,
+                    dr7 as *mut libc::c_void,
+                )?;
+                ptrace::ptrace(
+                    ptrace::Request::PTRACE_POKEUSER,
+                    self.pid,
+                    (*DEBUG_REG_OFFSET + 6 * 8) as *mut libc::c_void,
+                    0 as *mut libc::c_void,
+                )?;
+            }
+
+            self.hardware_breakpoints[index] = Some(breakpoint);
+
+            Ok(index)
+        }
+        #[cfg(not(target_arch = "x86_64"))]
+        Err(Box::new(HardwareBreakpointError::UnsupportedPlatform))
+    }
+
+    pub fn clear_hardware_breakpoint(
+        &mut self,
+        index: usize,
+    ) -> Result<HardwareBreakpoint, Box<dyn std::error::Error>> {
+        #[cfg(target_arch = "x86_64")]
+        {
+            if self.hardware_breakpoints[index].is_none() {
+                return Err(Box::new(HardwareBreakpointError::DoesNotExist(index)));
+            }
+
+            let mut dr7 =
+                self.ptrace_peekuser((*DEBUG_REG_OFFSET + 7 * 8) as *mut libc::c_void)? as u64;
+            let mut dr6 =
+                self.ptrace_peekuser((*DEBUG_REG_OFFSET + 6 * 8) as *mut libc::c_void)? as u64;
+
+            let dr7_bit_mask: u32 = (0b11 << (2 * index)) | (0b1111 << (16 + 4 * index));
+            dr7 = dr7 & !(dr7_bit_mask as u64);
+
+            let dr6_bit_mask: u32 = 1 << index;
+            dr6 = dr6 & !(dr6_bit_mask as u64);
+
+            #[allow(deprecated)]
+            unsafe {
+                //Have to use deprecated function because of no alternative for PTRACE_POKEUSER
+                ptrace::ptrace(
+                    ptrace::Request::PTRACE_POKEUSER,
+                    self.pid,
+                    (*DEBUG_REG_OFFSET + 7 * 8) as *mut libc::c_void,
+                    dr7 as *mut libc::c_void,
+                )?;
+                ptrace::ptrace(
+                    ptrace::Request::PTRACE_POKEUSER,
+                    self.pid,
+                    (*DEBUG_REG_OFFSET + 6 * 8) as *mut libc::c_void,
+                    dr6 as *mut libc::c_void,
+                )?;
+            }
+
+            let watchpoint = std::mem::replace(&mut self.hardware_breakpoints[index], None);
+            Ok(watchpoint.unwrap())
+        }
+
+        #[cfg(not(target_arch = "x86_64"))]
+        Err(Box::new(HardwareBreakpointError::UnsupportedPlatform))
+    }
+
+    pub fn clear_all_hardware_breakpoints(&mut self) -> Result<(), Box<dyn std::error::Error>> {
+        {
```

Comment:
I think we don't need this to be nested? I.e. this can probably just go into the `fn` scope